### PR TITLE
Add more unit tests and other cleanups.

### PR DESCRIPTION
**Pull request by haydenroche5**
- Fix some typos in n_cjson_helpers.c Doxygen comments.
- Check return value of _SerialReset in retry loop of serialNoteReset.
- Exclude third party code (e.g. n_md5.c) from coverage report.
- Remove unnecessary "inLog" code from NotePrint.
- Get rid of the need for some casting in i2cNoteReset and serialNoteTransaction.
- Add new unit tests for serialNoteTransaction, serialNoteReset, NoteSetFn, NoteDebug, NoteTransaction hooks, NoteSerial hooks, NotePrint, and NotePrintln.

### CODE REVIEW

File: n_serial.c [hunk 1]
```diff
@@ -48,7 +48,7 @@ const char *serialNoteTransaction(char *json, char **jsonResponse)
         if (segLen > CARD_REQUEST_SERIAL_SEGMENT_MAX_LEN) {
             segLen = CARD_REQUEST_SERIAL_SEGMENT_MAX_LEN;
         }
-        _SerialTransmit((uint8_t *)&transmitBuf[segOff], segLen, false);
+        _SerialTransmit(&transmitBuf[segOff], segLen, false);
```
Comment (zfields) [inline n_serial.c:51]: Based on the type of `transmitBuf` above, this looks like a nice clean-up. :+1:

But just to confirm, the tests running with `-Wall`, `-Wextra`, `-Wpedantic`, `-Werror`, right?

The reason I ask, is because a lot of the casting you see throughout the code was to keep different compilers from issuing warnings about our code. This one seems to be a non-issue, but it made me think about us having our flags enabled.

_I didn't look too hard, but I couldn't immediately find them the `CMakeLists.txt` below._

**Review by zfields — CHANGES_REQUESTED** [review]

**Comment by rayozzie** [comment]
It’s yet another hack-y way of getting debug output that is probably undocumented and you can probably desupport if you like.

Where it has come in useful is this:

  *   Someone develops an app with note-c (or note-arduino, or whatnot)
  *   Because of the nature of their hardware, they have no “Arduino debug console” or other such thing where they can see debug output on their mcu.  (We had this problem badly with, I believe, some little msp432 we were playing with.)  BUT they are desperate to see what is happening inside their host MCU
  *   The NoteDebug, when hooks aren’t registered, will call card.debug and output the app’s debug output on the notecarrier’s “trace mode” usb console
  *   The reason for “inLog” is self-evident if you were to remove it:  NoteTransaction ITSELF does NoteDebug output for transaction debugging, and without the flag we’d go recursive and blow the stack

From: Zachary J. Fields ***@***.***>
Date: Thursday, January 12, 2023 at 6:13 AM
To: blues/note-c ***@***.***>
Cc: Ray Ozzie ***@***.***>, Mention ***@***.***>
Subject: Re: [blues/note-c] Add more unit tests and other cleanups. (PR #38)

@zfields requested changes on this pull request.

________________________________

In n_helpers.c<https://github.com/blues/note-c/pull/38#discussion_r1068105717>:

> @@ -186,15 +186,9 @@ bool NotePrint(const char *text)



         NoteDebug(text);



         return true;



     }



-    int inLog = 0;



@rayozzie<https://github.com/rayozzie> , Can you remind us of the original intent of the inLog variable?

Obviously, it wasn't doing anything as it stood, but is there some functionality we should have and are missing?

________________________________

In n_i2c.c<https://github.com/blues/note-c/pull/38#discussion_r1068124854>:

> @@ -242,14 +242,14 @@ bool i2cNoteReset()



     for (retries=0; transmitErr==NULL && !notecardReady && retries<3; retries++) {







         // Loop to drain all chunks of data that may be ready to transmit to us



-        int chunklen = 0;



+        uint32_t chunklen = 0;


⬇️ Suggested change

-        uint32_t chunklen = 0;



+        uint16_t chunklen = 0;



I'm always leary of changing types.

Moving to unsigned is a good change, and I cannot see any immediate side-effects of uint32_t. However, since you are making the change, perhaps you can select a type that is "guaranteed to be safe", like uint16_t.

The signature for _I2CReceive() is:

const char *NoteI2CReceive(uint16_t DevAddress, uint8_t* pBuffer, uint16_t Size, uint32_t *available)

chunklen goes into the Size parameter, which is uint16_t.

As note-c written today, the value in chunklen cannot exceed 127, so there is no problem. That being said, why not play is safe?

________________________________

In n_serial.c<https://github.com/blues/note-c/pull/38#discussion_r1068148716>:

> @@ -212,8 +212,9 @@ bool serialNoteReset()



         _Debug("no notecard\n");



 #endif



         _DelayMs(500);



-        _SerialReset();



Whoa. Does this change behavior?

I'm concerned, because note-c is used by all our customers. If this is a bug, then it probably needs to be documented and stay a bug, because it's likely either depended upon or already been worked around, and the worst thing we could do is break our customers code.

@rayozzie<https://github.com/rayozzie> Can you bless this change?

________________________________

In n_serial.c<https://github.com/blues/note-c/pull/38#discussion_r1068159534>:

> @@ -48,7 +48,7 @@ const char *serialNoteTransaction(char *json, char **jsonResponse)



         if (segLen > CARD_REQUEST_SERIAL_SEGMENT_MAX_LEN) {



             segLen = CARD_REQUEST_SERIAL_SEGMENT_MAX_LEN;



         }



-        _SerialTransmit((uint8_t *)&transmitBuf[segOff], segLen, false);



+        _SerialTransmit(&transmitBuf[segOff], segLen, false);



Based on the type of transmitBuf above, this looks like a nice clean-up. 👍

But just to confirm, the tests running with -Wall, -Wextra, -Wpedantic, -Werror, right?

The reason I ask, is because a lot of the casting you see throughout the code was to keep different compilers from issuing warnings about our code. This one seems to be a non-issue, but it made me think about us having our flags enabled.

I didn't look too hard, but I couldn't immediately find them the CMakeLists.txt below.

________________________________

In n_i2c.c<https://github.com/blues/note-c/pull/38#discussion_r1068144640>:

> @@ -242,14 +242,14 @@ bool i2cNoteReset()



     for (retries=0; transmitErr==NULL && !notecardReady && retries<3; retries++) {







         // Loop to drain all chunks of data that may be ready to transmit to us



-        int chunklen = 0;



+        uint32_t chunklen = 0;



It's odd that the available parameter returns a uint32_t, which is then loaded into chunklen. 🤔

NoteI2cReceive is executing a callback (beyond our control), so it seems like we should program defensively and also add a bounds check (along with test) around this edge case.

Below (line #266) you could change the logic to protect against it with:

            // Read everything that's left on the module



            chunklen = (available > USHRT_MAX) ? USHRT_MAX : available;

File: n_serial.c (hunk 1 above)
Comment (haydenroche5) [reply n_serial.c:51]: It's in the top-level CMakeLists.txt:

```
add_compile_options(
    -Wall
    -Wextra
    -Wpedantic
    -Werror
)
```

File: n_serial.c [hunk 2]
```diff
@@ -212,8 +212,9 @@ bool serialNoteReset()
         _Debug("no notecard\n");
 #endif
         _DelayMs(500);
-        _SerialReset();
```
[the remark this answers is not in the document]
Comment (haydenroche5) [reply n_serial.c:215]: A couple years ago Ray changed the serial and I2C reset functions to return a bool:

```
commit 808dd1edca52e28b03fda10af0d59bd3dfccafb8
Author: Ray Ozzie <ray@ozzie.net>
Date:   Wed Aug 12 10:02:21 2020 -0400

    fix

diff --git a/n_lib.h b/n_lib.h
index a8cc84f..ac90f15 100644
--- a/n_lib.h
+++ b/n_lib.h
@@ -40,11 +40,11 @@ bool serialNoteReset(void);
 // Hooks
 void NoteLockNote(void);
 void NoteUnlockNote(void);
-void NoteSerialReset(void);
+bool NoteSerialReset(void);
 void NoteSerialTransmit(uint8_t *, size_t, bool);
 bool NoteSerialAvailable(void);
 char NoteSerialReceive(void);
-void NoteI2CReset(void);
+bool NoteI2CReset(void);
 const char *NoteI2CTransmit(uint16_t DevAddress, uint8_t* pBuffer, uint16_t Size);
 const char *NoteI2CReceive(uint16_t DevAddress, uint8_t* pBuffer, uint16_t Size, uint32_t *avail);
 bool NoteHardReset(void);
```

It seems to me that we wouldn't want to do the retry loop again if we're not able to reset the serial bus. Indeed, that's what's already happening at the start of `serialNoteReset`:

```
if (!_SerialReset()) {
    return false;
}
```

> because it's likely either depended upon or already been worked around

I think it's likely neither of those, actually. As I said, if we've made it this far `_SerialReset` already succeeded once. It seems unlikely to me that it would fail again here, so I would guess no one's actually hitting this in "real life." Also, if it did fail, I would think the bus is in an unusable state, so retrying would have no effect anyway, as transmit/receive aren't going to work.

File: n_serial.c (hunk 1 above)
Comment (zfields) [reply n_serial.c:51]: I thought I remembered you doing that, but I couldn't find it. :+1:

**Comment by haydenroche5** [comment]
> The reason for “inLog” is self-evident if you were to remove it:  NoteTransaction ITSELF does NoteDebug output for transaction debugging, and without the flag we’d go recursive and blow the stack

Correct, but the function in question is `NotePrint`, not `NoteDebug`. `NoteTransaction` should never call `NotePrint`, from my reading of the code.

**Comment by rayozzie** [comment]
get rid of the inLog then.  it must be historical before NoteDebug was separated from NotePrint

**Review by zfields — APPROVED** [review]
